README.md
# exchange

[![Go Report Card](https://goreportcard.com/badge/github.com/rehhouari/exchange)](https://goreportcard.com/report/github.com/rehhouari/exchange)

Go library for current & historical exchange rates, forex & cryptocurrency conversion, fluctuation, and timeseries
using the new [Free foreign exchange rates API](https://exchangerate.host/#/) by [arzzen](https://github.com/arzzen/)
([GitHub](https://github.com/arzzen/exchangerate.host))

## Features:

- Currency conversion, historical & current exchange rates, timeseries and fluctuations.
- No authentication/token needed!
- Select any base currency.
- 171 forex currency and 6000+ cryptocurrency!
- No dependencies, only standard library.
- Easy to use.

> **Upgrading from [asvvvad/exchange](https://github.com/asvvvad1/exchange) to version 1.0.0
> (for further upgrading please view [CHANGELOG.md](CHANGELOG.md))**:

- no built-in caching, do it yourself
- changed return types from `big.Float` to `float64`
  and the second argument's type for `ConvertTo()` from `int` to `float64`

## Usage:

> #### `go get -u github.com/rehhouari/exchange`

```go
package main

import (
	"fmt"

	"github.com/rehhouari/exchange"
)

func main() {
	// Create a new Exchange instance and set USD as the base currency for the exchange rates and conversion
	ex := exchange.New("USD")
	// convert 10 USD to EUR
	fmt.Println(ex.ConvertTo("EUR", 10.0))

	// You can convert between 171 fiat and +6000 cryptocurrency as well!
	// convert 10 USD to BTC
	fmt.Println(ex.ConvertTo("BTC", 10.0))
	// convert 10 USD to EUR at 2012-12-12 (date must be in the format YYYY-MM-DD)
	fmt.Println(ex.ConvertAt("2012-12-12", "EUR", 10.0))

	// Get the available forex/fiat codes ([]string)
	forexCodes, _ := ex.ForexCodes()

	// Get the available crypto codes ([]string)
	// Warning: +6000
	cryptoCodes, _ := ex.CryptoCodes()
	fmt.Println(cryptoCodes)

	// Get the forex codes data, includes code and description.
	forexData, _ := ex.ForexData()

	// Get the crypto codes data, includes code and description.
	// Warning: +6000
	cryptoData, _ := ex.CryptoData()
	fmt.Println(cryptoData["BTC"]["code"])

	// loop through the forex codes
	for _, code := range forexCodes {
		// print the forex codes data in the format: USD: US Dollar
		fmt.Println(code+":", forexData[code]["description"])
	}

	// Change the base currency to euro
	ex.SetBase("EUR")
	// Get the latest exchange rates with all currencies (Base is EUR)
	// exchangerate.host update them at midnight GMT so make sure to cache it till next day
	fmt.Println(ex.LatestRatesAll())

	// Get the latest rates with multiple currencies, not all (USD and JPY only)
	fmt.Println(ex.LatestRatesMultiple([]string{"USD", "JPY"}))

	// Get the exchange rates at 2012-12-12 but only with USD
	fmt.Println(ex.HistoricalRatesSingle("2012-12-12", "USD"))

	// Get historical rates between 2012 12 10 and 2012 12 12 for JPY and GBP
	fmt.Println(ex.TimeseriesMultiple("2012-12-10", "2012-12-12", []string{"USD", "JPY"}))

	// Get the fluctuation between 2012 12 10 and 2012 12 12 with USD
	fluctuation, _ := ex.FluctuationSingle("2012-12-10", "2012-12-12", "USD")
	// Print the change
	fmt.Println(fluctuation["change"])
}
```

### Results returned by each method:

- **ConvertTo**, **ConvertAt**, **HistoricalRatesSingle**, **LatestRatesSingle**
- - `float64`, error
- **LatestRatesAll**, **LatestRatesMultiple**, **HistoricalRatesAll**, **HistoricalRatesMultiple**:
- - `map[symbol(string)]rate(float64)`
- **ForexCodes**
- - `[]string{codes}`, error
- **ForexData**
- - `map[symbol]map[ code description ]string`, error
- **CryptoCodes**
- - `[]string{codes}`, error
- **CryptoData**
- - `map[symbol]map[ symbol name ]string`, error
- **FluctuationAll**, FluctuationMultiple,
- - `map[symbol]map[ start_rate end_rate change change_pct ]float64`, error
- **FluctuationSingle**
- - `map[ start_rate end_rate change change_pct ]float64`, error

- **TimeseriesAll**, **TimeseriesMultiple**
- - `map[date]map[symbols]float64`, error
- **TimeseriesSingle**
- - `map[date]map[symbol]float64`, error

## Notes:

- Exchange rates are refreshed every **midnight GMT**, cache results when using!
- You can use All, Multiple, Single with all of LatestRates, HistoricalRates, Timeseries and Fluctuation.
- Oldest date for historical rates and conversion is 1999-01-04
- Maximum allowed timeframe for Timeseries is 365 days
- Use [decimal](https://github.com/shopspring/decimal) to handle `float64` results with `decimal.NewFromFloat`.

#### Input validation with the appropriate errors for all methods is provided to help debug

#### Any help and contribution is welcome!

This is my first Go library and I had trouble with JSON parsing
(and I still do, didn't use `bitly/simplejson` to reduce dependencies).
There's a lot of room for improvement.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2022-10-20

### Added

- Added `func (exchange *Exchange) SetContext(context context.Context)` to allow passing of context to
  `net/http` package.
- Some tests added.

### Fixed

- Fixed validation of HTTP request query parameters.

## [1.0.0] - 2021-10-02

- Initial release.


example_test.go
package exchange

import (
	"fmt"
)

func ExampleExchange() {
	// Create a new Exchange instance and set USD as the base currency for the exchange rates and conversion
	ex := New("USD")
	// convert 10 USD to EUR
	fmt.Println(ex.ConvertTo("EUR", 10.0))

	// You can convert between 171 fiat and +6000 cryptocurrency as well!
	// convert 10 USD to BTC
	fmt.Println(ex.ConvertTo("BTC", 10.0))
	// convert 10 USD to EUR at 2012-12-12 (date must be in the format YYYY-MM-DD)
	fmt.Println(ex.ConvertAt("2012-12-12", "EUR", 10.0))

	// Get the available forex/fiat codes ([]string)
	forexCodes, _ := ex.ForexCodes()

	// Get the available crypto codes ([]string)
	// Warning: +6000
	cryptoCodes, _ := ex.CryptoCodes()
	fmt.Println(cryptoCodes)

	// Get the forex codes data, includes code and description.
	forexData, _ := ex.ForexData()

	// Get the crypto codes data, includes code and description.
	// Warning: +6000
	cryptoData, _ := ex.CryptoData()
	fmt.Println(cryptoData["BTC"]["code"])

	// loop through the forex codes
	for _, code := range forexCodes {
		// print the forex codes data in the format: USD: US Dollar
		fmt.Println(code+":", forexData[code]["description"])
	}

	// Change the base currency to euro
	ex.SetBase("EUR")
	// Get the latest exchange rates with all currencies (Base is EUR)
	// exchangerate.host update them at midnight GMT so make sure to cache it till next day
	fmt.Println(ex.LatestRatesAll())

	// Get the latest rates with multiple currencies, not all (USD and JPY only)
	fmt.Println(ex.LatestRatesMultiple([]string{"USD", "JPY"}))

	// Get the exchange rates at 2012-12-12 but only with USD
	fmt.Println(ex.HistoricalRatesSingle("2012-12-12", "USD"))

	// Get historical rates between 2012 12 10 and 2012 12 12 for JPY and GBP
	fmt.Println(ex.TimeseriesMultiple("2012-12-10", "2012-12-12", []string{"USD", "JPY"}))

	// Get the fluctuation between 2012 12 10 and 2012 12 12 with USD
	fluctuation, _ := ex.FluctuationSingle("2012-12-10", "2012-12-12", "USD")
	// Print the change
	fmt.Println(fluctuation["change"])
}


exchange_test.go
package exchange

import (
	"fmt"
	"testing"
)

func TestExchange(t *testing.T) {
	// Create a new Exchange instance and set USD as the base currency for the exchange rates and conversion
	ex := New("USD")
	// convert 10 USD to EUR
	fmt.Println(ex.ConvertTo("EUR", 10.0))

	// You can convert between 171 fiat and +6000 cryptocurrency as well!
	// convert 10 USD to BTC
	fmt.Println(ex.ConvertTo("BTC", 10.0))
	// convert 10 USD to EUR at 2012-12-12 (date must be in the format YYYY-MM-DD)
	fmt.Println(ex.ConvertAt("2012-12-12", "EUR", 10.0))

	// Get the available forex/fiat codes ([]string)
	forexCodes, _ := ex.ForexCodes()

	// Get the available crypto codes ([]string)
	// Warning: +6000
	cryptoCodes, _ := ex.CryptoCodes()
	fmt.Println(cryptoCodes)

	// Get the forex codes data, includes code and description.
	forexData, _ := ex.ForexData()

	// Get the crypto codes data, includes code and description.
	// Warning: +6000
	cryptoData, _ := ex.CryptoData()
	fmt.Println(cryptoData["BTC"]["code"])

	// loop through the forex codes
	for _, code := range forexCodes {
		// print the forex codes data in the format: USD: US Dollar
		fmt.Println(code+":", forexData[code]["description"])
	}

	// Change the base currency to euro
	ex.SetBase("EUR")
	// Get the latest exchange rates with all currencies (Base is EUR)
	// exchangerate.host update them at midnight GMT so make sure to cache it till next day
	fmt.Println(ex.LatestRatesAll())

	// Get the latest rates with multiple currencies, not all (USD and JPY only)
	fmt.Println(ex.LatestRatesMultiple([]string{"USD", "JPY"}))

	// Get the exchange rates at 2012-12-12 but only with USD
	fmt.Println(ex.HistoricalRatesSingle("2012-12-12", "USD"))

	// Get historical rates between 2012 12 10 and 2012 12 12 for JPY and GBP
	fmt.Println(ex.TimeseriesMultiple("2012-12-10", "2012-12-12", []string{"USD", "JPY"}))

	// Get the fluctuation between 2012 12 10 and 2012 12 12 with USD
	fluctuation, _ := ex.FluctuationSingle("2012-12-10", "2012-12-12", "USD")
	// Print the change
	fmt.Println(fluctuation["change"])
}


latest_rates_multiple_test.go
package exchange

import (
	"context"
	"net/url"
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_LatestRatesMultiple_OnBadCurrencyCode_ShouldReturnError(t *testing.T) {
	ex := New("USD")
	symbols := []string{"UNKNOWN"}

	rates, err := ex.LatestRatesMultiple(symbols)
	assert.IsType(t, map[string]float64{}, rates)
	assert.Len(t, rates, 0)
	assert.ErrorIs(t, err, ErrInvalidCode)
}

func Test_LatestRatesMultiple_ShouldReturnData(t *testing.T) {
	ex := New("USD")
	symbols := []string{"EUR", "JPY", "USD"}

	rates, err := ex.LatestRatesMultiple(symbols)
	assert.IsType(t, map[string]float64{}, rates)
	assert.Len(t, rates, len(symbols))
	assert.NoError(t, err)
}

func Test_LatestRatesMultiple_WithContext_ShouldReturnData(t *testing.T) {
	ctx, cancelFunction := context.WithCancel(context.Background())
	defer cancelFunction()

	ex := New("USD")
	ex.SetContext(ctx)
	symbols := []string{"EUR", "JPY", "USD"}

	rates, err := ex.LatestRatesMultiple(symbols)
	assert.IsType(t, map[string]float64{}, rates)
	assert.Len(t, rates, len(symbols))
	assert.NoError(t, err)
}

func Test_LatestRatesMultiple_WithCanceledContext_ShouldReturnError(t *testing.T) {
	ctx, cancelFunction := context.WithCancel(context.Background())
	cancelFunction()

	ex := New("USD")
	ex.SetContext(ctx)
	symbols := []string{"EUR", "JPY", "USD"}

	rates, err := ex.LatestRatesMultiple(symbols)
	assert.IsType(t, map[string]float64{}, rates)
	assert.Len(t, rates, 0)
	expectedErr := &url.Error{}
	assert.ErrorAs(t, err, &expectedErr)
}


exchange.go
package exchange

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// ErrInvalidCode is returned when the currency code is invalid
var ErrInvalidCode = errors.New("invalid currency code")

// ErrInvalidDate is returned when the date is too old
var ErrInvalidDate = errors.New("oldest possible date is 1999-01-04")

// ErrInvalidDateFormat is returned when the date isn't formatted as YYYY-MM-DD
var ErrInvalidDateFormat = errors.New("date format must be YYYY-MM-DD")

// ErrTimeframeExceeded is returned when the time between start_date and end_date is bigger than 365 days
var ErrTimeframeExceeded = errors.New("maximum allowed timeframe is 365 days")

// ErrInvalidTimeFrame is returned when the to date is older than to date. For example flipped the arguments
var ErrInvalidTimeFrame = errors.New("\"from\" date must be older than \"to\" date")

// ErrInvalidAPIResponse is returned when the API return success: false
var ErrInvalidAPIResponse = errors.New("unknown API error")

const (
	baseURL             string = "https://api.exchangerate.host"
	symbolsURL                 = baseURL + "/symbols"
	cryptocurrenciesURL        = baseURL + "/cryptocurrencies"
	latestURL                  = baseURL + "/latest"
	convertURL                 = baseURL + "/convert"
	historicalURL              = baseURL + "/"
	timeseriesURL              = baseURL + "/timeseries"
	fluctuationURL             = baseURL + "/fluctuation"
)

// Exchange is returned by New() and allows access to the methods
type Exchange struct {
	Base          string
	Context       context.Context
	isInitialized bool // is set to true if used via New
}

type query struct {
	From      string
	To        string
	Base      string
	Amount    float64
	Symbols   []string
	Date      string
	TimeFrame [2]string
}

var client = http.Client{}

// New creates a new instance of Exchange
func New(base string) *Exchange {
	x := &Exchange{
		Base:          base,
		isInitialized: true,
	}
	return x
}

// SetBase sets a new base currency for the exchange rates
func (exchange *Exchange) SetBase(base string) error {
	if err := ValidateCode(base); err != nil {
		return err
	}
	exchange.Base = base
	return nil
}

// SetContext sets a new context to be used in HTTP requests
func (exchange *Exchange) SetContext(context context.Context) {
	exchange.Context = context
}

// ValidateCode validates a single symbol code
func ValidateCode(code string) error {
	if len(code) != 3 {
		return ErrInvalidCode
	}
	return nil
}

// ValidateSymbols validates all symbols' codes in an array
func ValidateSymbols(symbols []string) error {
	for code := range symbols {
		if err := ValidateCode(symbols[code]); err != nil {
			return err
		}
	}

	return nil
}

// ValidateDate validates date string according to YYYY-MM-DD format and if it's
func ValidateDate(date string) error {
	matched, err := regexp.Match("[0-9]{4}-((0[1-9])|(1[0-2]))-([0-3][0-9])", []byte(date))
	if err != nil {
		return err
	}
	if !matched {
		return ErrInvalidDateFormat
	}
	oldestDate, _ := time.Parse("2006-01-02", "1999-01-03")
	selectedDate, _ := time.Parse("2006-01-02", date)
	if selectedDate.Before(oldestDate) {
		return ErrInvalidDate
	}
	return nil
}

// ValidateTimeFrame checks if the "from" and "to" date are not more than 365 days apart, and they're not mixed
func ValidateTimeFrame(TimeFrame [2]string) error {
	from, _ := time.Parse("2006-01-02", TimeFrame[0])
	to, _ := time.Parse("2006-01-02", TimeFrame[1])
	if to.Before(from) {
		return ErrInvalidTimeFrame
	}

	if to.Sub(from).Hours() > 8759.992992006 {
		return ErrTimeframeExceeded
	}

	return nil
}

func (exchange *Exchange) get(url string, q query) (map[string]interface{}, error) {
	var req *http.Request
	var err error

	if exchange.Context != nil {
		req, err = http.NewRequestWithContext(exchange.Context, "GET", url, nil)
	} else {
		req, err = http.NewRequest("GET", url, nil)
	}
	if err != nil {
		return nil, err
	}

	err = processQuery(req, q)
	if err != nil {
		return nil, err
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}

	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return nil, err
	}

	success := result["success"]
	if !success.(bool) {
		return nil, ErrInvalidAPIResponse
	}

	return result, nil
}

func addToQuery(req *http.Request, key string, value string) {
	q := req.URL.Query()          // Get a copy of the query values.
	q.Add(key, value)             // Add a new value to the set.
	req.URL.RawQuery = q.Encode() // Encode and assign back to the original query.
}

func processQuery(req *http.Request, q query) error {
	if q.Base != "" {
		if err := ValidateCode(q.Base); err != nil {
			return err
		}
		addToQuery(req, "base", q.Base)
	}

	if q.From != "" {
		if err := ValidateCode(q.From); err != nil {
			return err
		}
		addToQuery(req, "from", q.From)
	}

	if q.To != "" {
		if err := ValidateCode(q.To); err != nil {
			return err
		}
		addToQuery(req, "to", q.To)
	}

	if q.Amount > 1.0 {
		addToQuery(req, "amount", strconv.FormatFloat(q.Amount, 'f', -1, 64))
	}

	if len(q.Symbols) != 0 {
		if err := ValidateSymbols(q.Symbols); err != nil {
			return err
		}
		addToQuery(req, "symbols", strings.Join(q.Symbols, ","))
	}

	if q.Date != "" {
		if err := ValidateDate(q.Date); err != nil {
			return err
		}
		addToQuery(req, "date", q.Date)
	}

	if q.TimeFrame != [2]string{} {
		for i := 0; i < 1; i++ {
			if err := ValidateDate(q.TimeFrame[i]); err != nil {
				return err
			}
		}
		if err := ValidateTimeFrame(q.TimeFrame); err != nil {
			return err
		}
		addToQuery(req, "start_date", q.TimeFrame[0])
		addToQuery(req, "end_date", q.TimeFrame[1])
	}

	return nil
}

func (exchange *Exchange) apiSymbols() (map[string]map[string]string, error) {
	resp, err := exchange.get(symbolsURL, query{})
	if err != nil {
		return nil, err
	}
	result := make(map[string]map[string]string)
	for code, data := range resp["symbols"].(map[string]interface{}) {
		values := make(map[string]string)
		for name, value := range data.(map[string]interface{}) {
			values[name] = value.(string)
		}
		result[code] = values
	}
	return result, nil
}

func (exchange *Exchange) apiCryptocurrencies() (map[string]map[string]string, error) {
	resp, err := exchange.get(cryptocurrenciesURL, query{})
	if err != nil {
		return nil, err
	}
	result := make(map[string]map[string]string)
	for code, data := range resp["cryptocurrencies"].(map[string]interface{}) {
		values := make(map[string]string)
		for name, value := range data.(map[string]interface{}) {
			values[name] = value.(string)
		}
		result[code] = values
	}
	return result, nil
}

func (exchange *Exchange) apiLatest(q query) (map[string]float64, error) {
	resp, err := exchange.get(latestURL, q)
	if err != nil {
		return nil, err
	}
	result := resp["rates"].(map[string]interface{})
	rates := make(map[string]float64, len(result))
	for key := range result {
		rates[key] = result[key].(float64)
	}
	return rates, nil
}

func (exchange *Exchange) apiConvert(q query) (float64, error) {
	resp, err := exchange.get(convertURL, q)
	if err != nil {
		return 0.0, err
	}
	result := resp["result"].(float64)
	return result, nil
}

func (exchange *Exchange) apiHistorical(q query) (map[string]float64, error) {
	if err := ValidateDate(q.Date); err != nil {
		return nil, err
	}
	url := historicalURL + q.Date
	q.Date = ""
	resp, err := exchange.get(url, q)
	if err != nil {
		return nil, err
	}
	result := resp["rates"].(map[string]interface{})
	rates := make(map[string]float64, len(result))
	for key := range result {
		rates[key] = result[key].(float64)
	}
	return rates, nil
}

func (exchange *Exchange) apiTimeseriesAndFluctuation(url string, q query) (map[string]map[string]float64, error) {
	resp, err := exchange.get(url, q)
	if err != nil {
		return nil, err
	}
	result := make(map[string]map[string]float64)
	for date, rates := range resp["rates"].(map[string]interface{}) {
		ratemap := make(map[string]float64)
		for symbol, rate := range rates.(map[string]interface{}) {
			frate := rate.(float64)
			ratemap[symbol] = frate
			result[date] = ratemap
		}
	}
	return result, nil
}

// ForexCodes returns and array of supported forex/fiat currency codes
func (exchange *Exchange) ForexCodes() ([]string, error) {
	var codes []string

	result, err := exchange.apiSymbols()
	if err != nil {
		return nil, err
	}

	for k := range result {
		codes = append(codes, k)
	}

	sort.Strings(codes)
	return codes, nil
}

// ForexData returns a map of supported forex/fiat currencies data (code & description)
func (exchange *Exchange) ForexData() (map[string]map[string]string, error) {
	return exchange.apiSymbols()
}

// CryptoCodes returns and array of supported cryptocurrency codes
func (exchange *Exchange) CryptoCodes() ([]string, error) {
	var codes []string

	result, err := exchange.apiCryptocurrencies()
	if err != nil {
		return nil, err
	}

	for k := range result {
		codes = append(codes, k)
	}

	sort.Strings(codes)
	return codes, nil
}

// CryptoData returns a map of supported cryptocurrencies data (name and symbol)
func (exchange *Exchange) CryptoData() (map[string]map[string]string, error) {
	return exchange.apiCryptocurrencies()
}

// LatestRatesAll returns the latest exchange rates for all supported currencies
func (exchange *Exchange) LatestRatesAll() (map[string]float64, error) {
	return exchange.apiLatest(query{Base: exchange.Base})
}

// LatestRatesMultiple returns the latest exchange rates for multiple currencies
func (exchange *Exchange) LatestRatesMultiple(symbols []string) (map[string]float64, error) {
	return exchange.apiLatest(query{Base: exchange.Base, Symbols: symbols})

}

// LatestRatesSingle returns the latest exchange rates for a single currencies
func (exchange *Exchange) LatestRatesSingle(symbol string) (float64, error) {
	resp, err := exchange.apiLatest(query{Base: exchange.Base, Symbols: []string{symbol}})
	if err != nil {
		return 0.0, err
	}
	return resp[symbol], nil
}

// ConvertTo converts the amount from the exchange.Base currency to the target currency
func (exchange *Exchange) ConvertTo(target string, amount float64) (float64, error) {
	return exchange.apiConvert(query{From: exchange.Base, To: target, Amount: amount})
}

// ConvertAt converts the amount from the exchange.Base currency to the target currency
// at a selected historical date
func (exchange *Exchange) ConvertAt(date string, target string, amount float64) (float64, error) {
	return exchange.apiConvert(query{From: exchange.Base, To: target, Amount: amount, Date: date})
}

// HistoricalRatesAll returns the historical exchange rates for all supported currencies
func (exchange *Exchange) HistoricalRatesAll(date string) (map[string]float64, error) {
	return exchange.apiHistorical(query{Base: exchange.Base, Date: date})
}

// HistoricalRatesMultiple returns the historical exchange rates for multiple currencies
func (exchange *Exchange) HistoricalRatesMultiple(date string, symbols []string) (map[string]float64, error) {
	return exchange.apiHistorical(query{Base: exchange.Base, Symbols: symbols, Date: date})

}

// HistoricalRatesSingle returns the historical exchange rates for a single currency
func (exchange *Exchange) HistoricalRatesSingle(date string, symbol string) (float64, error) {
	resp, err := exchange.apiHistorical(query{Base: exchange.Base, Symbols: []string{symbol}, Date: date})
	if err != nil {
		return 0.0, err
	}
	return resp[symbol], nil
}

// TimeseriesAll returns the timeseries for all supported symbols
func (exchange *Exchange) TimeseriesAll(start string, end string) (map[string]map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(timeseriesURL, query{TimeFrame: [2]string{start, end}})
	return resp, err
}

// TimeseriesMultiple returns the timeseries for multiple symbols
func (exchange *Exchange) TimeseriesMultiple(start string, end string, symbols []string) (map[string]map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(timeseriesURL, query{TimeFrame: [2]string{start, end}, Symbols: symbols})
	return resp, err
}

// TimeseriesSingle returns the timeseries for a single symbol<
func (exchange *Exchange) TimeseriesSingle(start string, end string, symbol string) (map[string]map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(timeseriesURL, query{TimeFrame: [2]string{start, end}, Symbols: []string{symbol}})
	return resp, err
}

// FluctuationAll returns the fluctuation for all supported symbols
func (exchange *Exchange) FluctuationAll(start string, end string) (map[string]map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(fluctuationURL, query{TimeFrame: [2]string{start, end}})
	return resp, err
}

// FluctuationMultiple returns the fluctuation for multiple symbols
func (exchange *Exchange) FluctuationMultiple(start string, end string, symbols []string) (map[string]map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(fluctuationURL, query{TimeFrame: [2]string{start, end}, Symbols: symbols})
	return resp, err
}

// FluctuationSingle returns the fluctuation for a single symbol
func (exchange *Exchange) FluctuationSingle(start string, end string, symbol string) (map[string]float64, error) {
	resp, err := exchange.apiTimeseriesAndFluctuation(fluctuationURL, query{TimeFrame: [2]string{start, end}, Symbols: []string{symbol}})
	return resp[symbol], err
}
